compiler/dyno/lib/resolution/Resolver.cpp

```diff
@@ -1841,67 +1868,111 @@ void Resolver::exit(const uast::New* node) {
 }
 
 bool Resolver::enter(const For* loop) {
-  enterScope(loop);
 
   const AstNode* iterand = loop->iterand();
   iterand->traverse(*this);
-
   ResolvedExpression& iterandRE = byPostorder.byAst(iterand);
 
-  auto& MSC = iterandRE.mostSpecific();
-  bool isIter = MSC.isEmpty() == false &&
-                MSC.numBest() == 1 &&
-                MSC.only()->untyped()->kind() == uast::Function::Kind::ITER;
-
-  bool wasResolved = iterandRE.type().isUnknown() == false &&
-                     iterandRE.type().isErroneousType() == false;
-
-  QualifiedType idxType;
-
-  if (isIter) {
-    idxType = iterandRE.type();
-  } else if (wasResolved) {
-    //
-    // Resolve "iterand.these()"
-    //
-    std::vector<CallInfoActual> actuals;
-    actuals.push_back(CallInfoActual(iterandRE.type(), USTR("this")));
-    auto ci = CallInfo (/* name */ USTR("these"),
-                        /* calledType */ iterandRE.type(),
-                        /* isMethod */ true,
-                        /* hasQuestionArg */ false,
-                        /* isParenless */ false,
-                        actuals);
-    auto inScope = scopeStack.back();
-    auto c = resolveGeneratedCall(context, iterand, ci, inScope, poiScope);
-
-    if (c.mostSpecific().only() != nullptr) {
-      idxType = c.exprType();
-      handleResolvedAssociatedCall(iterandRE, loop, ci, c);
+  if (loop->isParam()) {
+    if (iterand->isRange() == false) {
+      context->error(loop, "param loops may only iterate over range literals");
+    } else {
+      // TODO: ranges with strides, '#', and '<'
+      const Range* rng = iterand->toRange();
+      ResolvedExpression& lowRE = byPostorder.byAst(rng->lowerBound());
+      ResolvedExpression& hiRE = byPostorder.byAst(rng->upperBound());
+      auto low = lowRE.type().param()->toIntParam();
+      auto hi = hiRE.type().param()->toIntParam();
+
+      if (low == nullptr || hi == nullptr) {
+        context->error(loop, "param loops may only iterate over range literals with integer bounds");
+      }
+
+      std::vector<ResolutionResultByPostorderID> loopResults;
+      for (int64_t i = low->value(); i <= hi->value(); i++) {
+        ResolutionResultByPostorderID bodyResults;
+        auto cur = Resolver::paramLoopResolver(*this, loop, bodyResults);
+
+        cur.enterScope(loop);
+
+        ResolvedExpression& idx = cur.byPostorder.byAst(loop->index());
+        QualifiedType qt = QualifiedType(QualifiedType::PARAM, lowRE.type().type(), IntParam::get(context, i));
+        idx.setType(qt);
+        loop->body()->traverse(cur);
+
+        cur.exitScope(loop);
+
+        loopResults.push_back(cur.byPostorder);
```

Comment:
This might do an expensive copy since byPostorder might be a vector and all that. The main way I know of to avoid the copy is `loopResults.push_back(ResolutionResultByPostorderID()); loopResults.back().swap(cur.byPostorder)`. But you might know of a better way.

Reply:
Will `std::move`'ing the `byPostorder` not work? `push_back` has an rvalue reference overload.

Reply:
Seems like a good plan if it would work! I keep thinking that `push_back` only takes in a reference so it wouldn't work but I'm always forgetting / confusing myself about rvalue references.